lib/confluence/failure_to_ship.es6.js

```diff
@@ -0,0 +1,136 @@
+          daoType: 'MDAO',
+        });
+      },
+    },
+  ],
+  methods: [
+    {
+      name: 'compute',
+      documentation: `Compute failure to ship value for each browser in
+           browsers at the given date.`,
+      args: [
+        {
+          name: 'browsers',
+          typeName: 'Array of Browser',
+          documentation: `An array of brwsers. These browser are either
+              released at given date, or it is the most recent release
+              before given date.`,
+        },
+        {
+          name: 'date',
+          typeName: 'Date',
+          documentation: `The failure to ship value will be calculated at
+              given date.`,
+        },
+      ],
+      code: function(browsers, date) {
+        for (let i = 0; i < browsers.length; i++) {
+          let browser = browsers[i];
+          let year = date.getFullYear();
+          let month = date.getMonth();
+          let day = date.getDate();
+          let previousYear = new Date(year - 1, month, day);
+          let browserKeys = browsers.map((browser) => browser.browserKey);
+          // Find all browsers that
+          //   - a previous version of the browser.
+          //   - some other browser but released in previous one year range.
+          //   - if no release in one year, uese most recent browser.
+          this.browserDAO.where(this.mlang.OR(
+            this.mlang.IN(this.Browser.BROWSER_KEY, browserKeys),
+            this.mlang.AND(
+              this.mlang.EQ(this.Browser.BROWSER_NAME, browser.browserName),
+              this.mlang.EQ(this.Browser.OS_NAME, browser.osName),
+              this.mlang.GT(this.Browser.RELEASE_DATE, previousYear),
+              this.mlang.LT(this.Browser.RELEASE_DATE, date)),
+            this.mlang.AND(
+              this.mlang.GT(this.Browser.RELEASE_DATE, previousYear),
+              this.mlang.LT(this.Browser.RELEASE_DATE, date))))
+          .select().then((result) => {
+            let browserIface = {};
+            let numOtherBrowsers = 0;
+            let majorIface = {};
+            let promises = [];
+            let prevReleaseBrowsers = [];
+            let comparedBrowsers = [];
+            for (let i = 0; i < result.a.length; i++) {
+              if (result.a[i].browserName !== browser.browserName) {
+                numOtherBrowsers++;
+                comparedBrowsers.push(result.a[i]);
+              } else if (result.a[i].browserKey !== browser.browserKey) {
+                prevReleaseBrowsers.push(result.a[i]);
+              }
+              promises.push(result.a[i].interfaces.select({
+                put: function(iface) {
+                  if (result.a[i].browserName === browser.browserName) {
+                    if (!browserIface.hasOwnProperty(iface.interfaceKey)) {
+                      browserIface[iface.interfaceKey] = true;
+                    }
+                  } else {
+                    if (!majorIface.hasOwnProperty(iface.interfaceKey)) {
```

Comment:
Add a comment here: Interfaces may have names defined in `majorIface`'s prototype chain; that's why we don't use `majorIface[iface.interfaceKey] = (majorIface[iface.interfaceKey] || 0) + 1`.